RESTWS-817: ConceptResource to GET concepts of a given class

josephbate opened the pull request:
<!--- Add a pull request title above in this format -->
<!--- real example: 'RESTWS-738 Remove concept property setters from ObsResource classes' -->
<!--- 'RESTWS-JiraIssueNumber JiraIssueTitle' -->
## Description of what I changed
1. added the name parameter to the optional parameters
2. added a method to check whether the concept class parameter is not null to invoke conceptservice.getConceptsByClass method
3. Addeda test to the ConceptSearchHandler1_9test.java to test the changes made 
<!--- Describe your changes in detail -->
<!--- It can simply be your commit message, which you must have -->


## Issue I worked on
<!--- This project only accepts pull requests related to open issues -->
<!--- Want a new feature or change? Discuss it in an issue first -->
<!--- Found a bug? Point us to the issue/or create one so we can reproduce it -->
<!--- Just add the issue number at the end: -->
see https://issues.openmrs.org/browse/RESTWS-817

## Checklist: I completed these to help reviewers :)
<!--- Put an `x` in the box if you did the task -->
<!--- If you forgot a task please follow the instructions below -->
- [ ] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

josephbate commented:
i have redone this issue for a more cleaner review and merge
please review this pullrequest
cc. @mozzy11 @r-nathan @ibacher @dkayiwa

dkayiwa commented on omod-1.8/src/main/java/org/openmrs/module/webservices/rest/web/v1_0/search/openmrs1_8/ConceptSearchHandler1_8.java:
```diff
@@ -162,6 +162,18 @@ public PageableResult search(RequestContext context) throws ResponseException {
+
+				List<Concept> concept = conceptService.getConceptsByClass(responseConceptClass);
+				concepts.addAll(concept);
+				return new NeedsPaging<Concept>(concepts, context);
```
The ticket description says that we should be able to filter concepts by class, in addition to other parameters. What you have implemented filters by class and leaves out the rest of the parameters. In other words, `AND` the given parameters. e.g `WHERE name=ASP && searchType=fuzzy && class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f`. Does this make sense?

josephbate replied:
> The ticket description says that we should be able to filter concepts by class, in addition to other parameters. What you have implemented filters by class and leaves out the rest of the parameters. In other words, `AND` the given parameters. e.g `WHERE name=ASP AND searchType=fuzzy AND class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f`. Does this make sense?

hello @dkayiwa i may not have gotten your comment correctly but i have just test this endpoint `/openmrs/ws/rest/v1/concept?name=ASP&searchType=fuzzy&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` and  it works fine unless i missed something in your comment please enlighted me 
 
because according to @mks-d  this endpont `/openmrs/ws/rest/v1/concept?class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` was the issue, that if the name parameter was dropped the class parameter would not work so i included a method that filters the concept 

are you suggesting that i include more  method to cater for the name as well?
please let me know

josephbate replied:
because i think thats the only one missing 
for in my understnding the searchType is already catered for with this method 
`// If there's class parameter in query
		if ("fuzzy".equals(searchType)) {
			List<Locale> locales = new ArrayList<Locale>(LocaleUtility.getLocalesInOrder());
			List<ConceptClass> classes = null;
			ConceptClass responseConceptClass = conceptService.getConceptClassByUuid(conceptClass);
			
			if (responseConceptClass != null) {
				classes = Arrays.asList(responseConceptClass);
			}
			
			List<ConceptSearchResult> searchResults = conceptService.getConcepts(name, locales, context.getIncludeAll(),
			    classes, null, null, null, null, context.getStartIndex(), context.getLimit());
			List<Concept> results = new ArrayList<Concept>(searchResults.size());
			for (ConceptSearchResult csr : searchResults) {
				results.add(csr.getConcept());
			}
			return new NeedsPaging<Concept>(results, context);
		} else if (searchType == null || "equals".equals(searchType)) {
			
			if (name != null) {
				Concept concept = conceptService.getConceptByName(name);
				concepts.add(concept);
				if (concept != null) {
					boolean isPreferredOrFullySpecified = false;
					for (ConceptName conceptname : concept.getNames()) {
						if (conceptname.getName().equalsIgnoreCase(name)
						        && (conceptname.isPreferred() || conceptname.isFullySpecifiedName())) {
							isPreferredOrFullySpecified = true;
							break;
						}
					}
					if (!isPreferredOrFullySpecified) {
						throw new APIException(
						        "The concept name should be either a fully specified or locale preferred name");
					}
					
					return new NeedsPaging<Concept>(concepts, context);
				} else {
					return new EmptySearchResult();
				}
			}
		} else {
			throw new InvalidSearchException("Invalid searchType: " + searchType
			        + ". Allowed values: \"equals\" and \"fuzzy\"");
		}`

josephbate replied:
plus if a drop the class parameter and just run http://localhost:8080/openmrs/ws/rest/v1/concept?name=ASP&searchType=fuzzy it still works

dkayiwa replied:
According to your implementation, if the class `8d490dfc-c2cc-11de-8d13-0010c6dffd0f` has concepts with names `ASP` and `XYZ`, it will return both of them simply because they have that class. Yet you are required to filter even the name within that class, according to the search parameters.

josephbate replied:
if am to get you correctly is that i should i write a method for that filters concepts by name such that if we are to call `/openmrs/ws/rest/v1/concept?name=ASP&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` it should filter by name and class, and also filter comcepts by only name also

is this so @dkayiwa

josephbate replied:
that is to say `/openmrs/ws/rest/v1/concept?name=ASP`

josephbate replied:
or you would want me to append the current implementation to add the name parameter as a filter

dkayiwa replied:
Let me put it differently, can you write three tests where one makes the call `/openmrs/ws/rest/v1/concept?name=ASP&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` and the other `/openmrs/ws/rest/v1/concept?name=XYZ&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f`, then lastly `/openmrs/ws/rest/v1/concept?class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f`

josephbate replied:
i got it now
let me try to make some changes

josephbate replied:
hello @dkayiwa 
i have managed to make some changes 
we can now search concept by both name and class or class alone 
however iam faced by a challenge it can only work if you spell out the fullnames of the concept that is ` /openmrs/ws/rest/v1/concept?name=caffeine&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` but abbreviations or call them acro-names ` /openmrs/ws/rest/v1/concept?name=CAF&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` to make a request is challenging

i kindly request further guidance on this 
thank

josephbate replied:
hello @dkayiwa 
as i explored further i release we maybe dont really need to write a new method for the classfilter logic we may just need to edit `if ("fuzzy".equals(searchType))` to something like this `if ("fuzzy".equals(searchType) || conceptClass != null)` becuase the code needed is the same and it corrects all the failures of the ticket 

However we run into a challage of testing the if else part where the results  should return an empty string 
i have commented out my implementation for now and the two test failures then complied the code ran sdk and tested out everything, it worked fine except for the empty result and the exception

josephbate replied:
i still for your comment @dkayiwa

josephbate replied:
what would be your on this please

josephbate commented:
Hello @dkayiwa  i have added the test case for both `/openmrs/ws/rest/v1/concept?name=ASP&class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f` and `/openmrs/ws/rest/v1/concept?class=8d490dfc-c2cc-11de-8d13-0010c6dffd0f

` 
but according to the tests there only to concept with the same name same uuid  i dont know how but to prove this i added `System.out.println(results);` to print then out

josephbate commented:
i realease if we add `|| conceptClass != null` to the if clause will not detect errors or validate the existance of the conceptclass which in turn cause leakage 
so am trying to work around the previous implementation so as to it allows both fully specified concept name entries and not fully specified
